### PR TITLE
[FLINK-12366][table] Clean up Catalog APIs to make them more consiste…

### PR DESCRIPTION
**Pull request by xuefuz**
…nt and coherent

## What is the purpose of the change

Clean up existing catalog APIs to make them more consistent and coherent

## Brief change log

  - Clean up API signature regarding exception thrown
  - Improve error handling
 
## Verifying this change

This change is already covered by existing tests. Tests are modified to reflect the proposed changes.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (avaDocs)

### CODE REVIEW

File: flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/catalog/GenericInMemoryCatalogTest.java [hunk 1]
```diff
@@ -731,32 +727,36 @@ public void testDropPartition() throws Exception {
...
+		assertEquals(Arrays.asList(createPartitionSpec()), catalog.listAllPartitions(path1));
 
 		catalog.dropPartition(path1, createPartitionSpec(), false);
 
-		assertEquals(Arrays.asList(), catalog.listPartitions(path1));
+		assertEquals(Arrays.asList(), catalog.listAllPartitions(path1));
 	}
 
 	@Test
 	public void testDropPartition_TableNotExistException() throws Exception {
```
Comment (bowenli86) [inline flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/catalog/GenericInMemoryCatalogTest.java]: We need to update the unit test names to reflect what case the test is actually about. E.g. this UT can be renamed to something like `testDropParition_tableNotExist()` or `testDropParition_ParitionNotExistException_tableNotExist()`.

Same thing for the following changed unit tests in this file.

**Comment by xuefuz** [comment]
Thanks, Bowen! Updated the PR to address above your review feedback.

**Comment by bowenli86** [comment]
@KurtYoung Would be good to have you take a look, and help to merge if there's no more concerns. Thanks!

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/ReadableCatalog.java [hunk 2]
```diff
@@ -174,14 +172,11 @@
...
 	 *
-	 * @throws TableNotExistException thrown if the table does not exist in the catalog
-	 * @throws TableNotPartitionedException thrown if the table is not partitioned
-	 * @throws PartitionSpecInvalidException thrown if the given partition spec is invalid,
 	 * @throws PartitionNotExistException thrown if the partition is not partitioned
 	 * @throws CatalogException	in case of any runtime exception
 	 */
 	CatalogPartition getPartition(ObjectPath tablePath, CatalogPartitionSpec partitionSpec)
-		throws TableNotExistException, TableNotPartitionedException, PartitionSpecInvalidException, PartitionNotExistException, CatalogException;
+		throws PartitionNotExistException, CatalogException;
```
Comment (KurtYoung) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/ReadableCatalog.java]: I don't quite get the idea why this function will not throw `TableNotExistException`, but `listPartitions` will. These two methods are both some kind of "get some partition from table".

Comment (xuefuz) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/ReadableCatalog.java]: In short, partition not existing includes the case of table not existing. getPartition() attempts to get a specific partition for a table. If the table doesn't exist, then of course, the partition doesn't exist.

On the other hand, listPartition() fails (throws an exception) if the table doesn't exist, which is reasonable.

Just like getTable(), which doesn't throw databasenotexistEx when db doesn't exist. The change makes the behavior consistent across board.

**Comment by KurtYoung** [comment]
LGTM, merging this

**Pull request merged**